File: py/src/infrastructure/repositories/in_memory_wallet_repository.py
import time
from src.domain.entities.wallet import Wallet
from src.domain.repositories.wallet_repository import WalletRepository
from src.domain.value_objects.currency import Currency
from src.domain.value_objects.money import Money

class ConcurrencyError(Exception):
    pass

class WalletRecord:
    def __init__(self, wallet_id: str, amount: float, currency_code: str, version: int):
        self.id = wallet_id
        self.amount = amount
        self.currency_code = currency_code
        self.version = version

class InMemoryWalletRepository(WalletRepository):
    def __init__(self, use_optimistic_locking: bool = True, simulated_delay_seconds: float = 0.0):
        self._records: dict[str, WalletRecord] = {}
        self.use_optimistic_locking = use_optimistic_locking
        self.simulated_delay_seconds = simulated_delay_seconds

    def find_by_id(self, wallet_id: str) -> Wallet | None:
        record = self._records.get(wallet_id)
        if not record:
            return None
        
        currency = Currency(record.currency_code)
        balance = Money(record.amount, currency)
        return Wallet(record.id, balance, record.version)

    def save(self, wallet: Wallet) -> None:
        # 1. Simulate DB latency
        if self.simulated_delay_seconds > 0.0:
            time.sleep(self.simulated_delay_seconds)

        current_record = self._records.get(wallet.id)
        if not current_record:
            # First time save
            self._records[wallet.id] = WalletRecord(
                wallet_id=wallet.id,
                amount=wallet.balance.amount,
                currency_code=wallet.balance.currency.code,
                version=wallet.version
            )
            return

        # 2. Check version for optimistic locking
        if self.use_optimistic_locking:
            if current_record.version != wallet.version:
                raise ConcurrencyError(
                    f"Optimistic locking conflict: Wallet {wallet.id} was updated by another request. "
                    f"Current DB version: {current_record.version}, wallet version: {wallet.version}"
                )
            
            # Successful write: increment version
            new_version = current_record.version + 1
            current_record.amount = wallet.balance.amount
            current_record.currency_code = wallet.balance.currency.code
            current_record.version = new_version
            
            # Synchronize the aggregate root in-memory version
            wallet.increment_version()
        else:
            # Overwrite without checks
            current_record.amount = wallet.balance.amount
            current_record.currency_code = wallet.balance.currency.code

    def seed(self, wallet_id: str, amount: float, currency_code: str) -> None:
        self._records[wallet_id] = WalletRecord(
            wallet_id=wallet_id,
            amount=amount,
            currency_code=currency_code,
            version=0
        )

    def clear(self) -> None:
        self._records.clear()

File: js/src/infrastructure/repositories/InMemoryWalletRepository.ts
import { Wallet } from "../../domain/entities/Wallet";
import { IWalletRepository } from "../../domain/repositories/IWalletRepository";
import { Currency } from "../../domain/value-objects/Currency";
import { Money } from "../../domain/value-objects/Money";

export class ConcurrencyException extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ConcurrencyException";
  }
}

interface WalletRecord {
  id: string;
  amount: number;
  currencyCode: string;
  version: number;
}

export class InMemoryWalletRepository implements IWalletRepository {
  private records = new Map<string, WalletRecord>();
  
  constructor(
    public useOptimisticLocking: boolean = true,
    public simulatedDelayMs: number = 0
  ) {}

  public async findById(id: string): Promise<Wallet | null> {
    const record = this.records.get(id);
    if (!record) return null;

    const currency = Currency.fromCode(record.currencyCode);
    const balance = new Money(record.amount, currency);
    
    // Reconstruct Wallet aggregate root with its current version
    return new Wallet(record.id, balance, record.version);
  }

  public async save(wallet: Wallet): Promise<void> {
    // 1. Simulate DB query/write latency
    if (this.simulatedDelayMs > 0) {
      await new Promise((resolve) => setTimeout(resolve, this.simulatedDelayMs));
    }

    const currentRecord = this.records.get(wallet.id);
    if (!currentRecord) {
      // If it's a new wallet, just insert it
      this.records.set(wallet.id, {
        id: wallet.id,
        amount: wallet.balance.amount,
        currencyCode: wallet.balance.currency.code,
        version: wallet.version,
      });
      return;
    }

    // 2. Perform Optimistic Locking check if enabled
    if (this.useOptimisticLocking) {
      if (currentRecord.version !== wallet.version) {
        throw new ConcurrencyException(
          `Optimistic locking conflict: Wallet ${wallet.id} was updated by another request. Current DB version: ${currentRecord.version}, wallet version: ${wallet.version}`
        );
      }
      
      // Successful update: increment database version
      const newVersion = currentRecord.version + 1;
      this.records.set(wallet.id, {
        id: wallet.id,
        amount: wallet.balance.amount,
        currencyCode: wallet.balance.currency.code,
        version: newVersion,
      });
      
      // Update in-memory aggregate version to stay in sync
      wallet.incrementVersion();
    } else {
      // Save without optimistic locking (overwriting version)
      this.records.set(wallet.id, {
        id: wallet.id,
        amount: wallet.balance.amount,
        currencyCode: wallet.balance.currency.code,
        version: currentRecord.version, // keep same version (or update it, but ignore checks)
      });
    }
  }

  /**
   * Helper to seed initial wallet data.
   */
  public seed(id: string, amount: number, currencyCode: string): void {
    this.records.set(id, {
      id,
      amount,
      currencyCode,
      version: 0,
    });
  }

  public clear(): void {
    this.records.clear();
  }
}
